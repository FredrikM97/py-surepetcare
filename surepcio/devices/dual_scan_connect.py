import logging
from datetime import time
from typing import Optional

from .device import BaseControl
from .device import BaseStatus
from .device import DeviceBase
from surepcio.command import Command
from surepcio.const import API_ENDPOINT_PRODUCTION
from surepcio.const import API_ENDPOINT_V1
from surepcio.devices.entities import DevicePetTag
from surepcio.entities.error_mixin import ImprovedErrorMixin
from surepcio.enums import ProductId

logger = logging.getLogger(__name__)


class Curfew(ImprovedErrorMixin):
    enabled: bool
    lock_time: time
    unlock_time: time


class Locking(ImprovedErrorMixin):
    mode: int = 0


class Control(BaseControl):
    curfew: Optional[list[Curfew]] = None
    locking: Optional[int] = None
    fail_safe: Optional[int] = None
    fast_polling: Optional[bool] = None


class Status(BaseStatus):
    locking: Optional[Locking] = None


class DualScanConnect(DeviceBase):
    controlCls = Control
    statusCls = Status

    @property
    def product(self) -> ProductId:
        return ProductId.DUAL_SCAN_CONNECT

    def refresh(self):
        def parse(response):
            if not response:
                return self
            self.status = Status(**{**self.status.model_dump(), **response["data"]})
            self.control = Control(**{**self.control.model_dump(), **response["data"]})
            self.tags = [DevicePetTag(**tag) for tag in response["data"].get("tags", [])]
            return self

        return Command(
            method="GET",
            endpoint=f"{API_ENDPOINT_PRODUCTION}/device/{self.id}",
            callback=parse,
        )

    def set_curfew(self, curfew: list[Curfew]) -> Command:
        """Set the flap curfew times, using the household's timezone"""

        def parse(response):
            if not response:
                return self
            # Unclear what to do with the data.. Should we refresh or is there any callback info?
            logger.info("Parse callback from curfew on device")
            return self

        return Command(
            "PUT",
            f"{API_ENDPOINT_V1}/device/{self.id}/control",
            params=Control(curfew=curfew).model_dump(),
            callback=parse,
        )

    def set_locking(self, locking: int) -> Command:
        """Set locking mode"""

        def parse(response):
            if not response:
                return self
            # Unclear what to do with the data.. Should we refresh or is there any callback info?
            logger.info("Parse callback from locking on device")
            return self

        return Command(
            "PUT",
            f"{API_ENDPOINT_V1}/device/{self.id}/control",
            params=Control(locking=locking).model_dump(),
            callback=parse,
        )

    def set_failsafe(self, failsafe: int) -> Command:
        """Set failsafe mode"""

        def parse(response):
            if not response:
                return self
            # Unclear what to do with the data.. Should we refresh or is there any callback info?
            logger.info("Parse callback from failsafe on device")
            return self

        return Command(
            "PUT",
            f"{API_ENDPOINT_V1}/device/{self.id}/control",
            params=Control(fail_safe=failsafe).model_dump(),
            callback=parse,
        )
