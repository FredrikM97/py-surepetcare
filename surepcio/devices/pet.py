import logging
from datetime import datetime
from datetime import timezone
from typing import Optional
from zoneinfo import ZoneInfo

from pydantic import Field
from pydantic import field_validator

from .device import PetBase
from surepcio.command import Command
from surepcio.const import API_ENDPOINT_PRODUCTION
from surepcio.entities.error_mixin import ImprovedErrorMixin
from surepcio.enums import BowlPosition
from surepcio.enums import FoodType
from surepcio.enums import ProductId

logger = logging.getLogger(__name__)


class ReportHouseholdMovementResource(ImprovedErrorMixin):
    """Represents a movement resource in the household report."""

    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    deleted_at: Optional[datetime] = None
    device_id: Optional[int] = None
    tag_id: Optional[int] = None
    user_id: Optional[int] = None
    from_: Optional[datetime] = Field(default=None, alias="from")
    to: Optional[datetime] = None
    duration: Optional[int] = None
    entry_device_id: Optional[int] = None
    entry_user_id: Optional[int] = None
    exit_device_id: Optional[int] = None
    exit_user_id: Optional[int] = None
    active: Optional[bool] = None
    exit_movement_id: Optional[int] = None
    entry_movement_id: Optional[int] = None


class ReportWeightFrame(ImprovedErrorMixin):
    """Represents a weight frame in the household report."""

    index: Optional[BowlPosition] = None
    weight: Optional[float] = None
    change: Optional[float] = None
    food_type_id: Optional[FoodType] = None
    target_weight: Optional[float] = None


class ReportHouseholdFeedingResource(ImprovedErrorMixin):
    """Represents a feeding resource in the household report."""

    from_: datetime = Field(alias="from")
    to: datetime
    duration: int
    context: int
    bowl_count: int
    device_id: int
    weights: list[ReportWeightFrame] = Field(default_factory=list)


class ReportHouseholdDrinkingResource(ImprovedErrorMixin):
    """Represents a drinking resource in the household report."""

    from_: Optional[datetime] = Field(default=None, alias="from")
    to: Optional[datetime] = None
    duration: Optional[int] = None
    context: Optional[str] = None
    bowl_count: Optional[int] = None
    device_id: Optional[int] = None
    weights: Optional[list[float]] = None
    actual_weight: Optional[float] = None
    entry_user_id: Optional[int] = None
    exit_user_id: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    deleted_at: Optional[datetime] = None
    tag_id: Optional[int] = None
    user_id: Optional[int] = None


class ReportHouseholdResource(ImprovedErrorMixin):
    movement: list[ReportHouseholdMovementResource] = Field(default_factory=list)
    feeding: list[ReportHouseholdFeedingResource] = Field(default_factory=list)
    drinking: list[ReportHouseholdDrinkingResource] = Field(default_factory=list)

    @field_validator("movement", "feeding", "drinking", mode="before")
    def extract_datapoints(v):
        if isinstance(v, dict) and "datapoints" in v:
            return v["datapoints"]
        return v


class Control(ImprovedErrorMixin):
    pass


class Status(ImprovedErrorMixin):
    report: ReportHouseholdResource = Field(default_factory=ReportHouseholdResource)


class Pet(PetBase[Control, Status]):
    controlCls = Control
    statusCls = Status

    def __init__(self, data: dict, **kwargs) -> None:
        super().__init__(data, **kwargs)
        # DateTime from user datetime to UTC timezone
        self.last_fetched_datetime: datetime = datetime.now(ZoneInfo(self.timezone)).astimezone(timezone.utc)

    @property
    def available(self) -> bool:
        """Static until figured out how to handle pet availability."""
        return True

    @property
    def photo(self) -> str | None:
        if self.entity_info.photo is None:
            return None
        return self.entity_info.photo.location

    def refresh(self) -> Command:
        """Refresh the pet's report data."""
        return self.fetch_report()

    def fetch_report(
        self,
        from_date: datetime | None = None,
        to_date: datetime | None = None,
        event_type: int | None = None,
    ) -> Command:
        def parse(response):
            self.status = Status(
                **{**self.status.model_dump(), "report": ReportHouseholdResource(**response["data"])}
            )
            self.control = Control(**{**self.control.model_dump(), **response["data"]})
            self.last_fetched_datetime = datetime.now(ZoneInfo(self.timezone))
            return self

        params = {}

        if not from_date:
            from_date = self.last_fetched_datetime
        params["From"] = from_date.isoformat()

        if not to_date:
            to_date = datetime.now(ZoneInfo(self.timezone)).astimezone(timezone.utc)
        params["To"] = to_date.isoformat()

        if event_type is not None:
            if event_type not in [1, 2, 3]:
                raise ValueError("event_type can only contain 1, 2, or 3")
            params["EventType"] = str(event_type)
        return Command(
            method="GET",
            endpoint=(
                f"{API_ENDPOINT_PRODUCTION}/report/household/{self.household_id}/pet/{self.id}/aggregate"
            ),
            params=params,
            callback=parse,
        )

    @property
    def product(self) -> ProductId:
        return ProductId.PET

    @property
    def tag(self) -> int | None:
        if self.entity_info.tag is None:
            logger.warning("Pet tag is not set")
            return None
        return self.entity_info.tag.id

    @property
    def feeding(self) -> list[ReportHouseholdFeedingResource]:
        return self.status.report.feeding

    @property
    def movement(self) -> list[ReportHouseholdMovementResource]:
        return self.status.report.movement

    @property
    def drinking(self) -> list[ReportHouseholdDrinkingResource]:
        return self.status.report.drinking
