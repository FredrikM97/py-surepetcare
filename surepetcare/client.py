import logging

from surepetcare.command import Command
from surepetcare.entities.api import ApiResponse
from surepetcare.security.auth import AuthClient

logger = logging.getLogger(__name__)


class SurePetcareClient(AuthClient, ApiResponse):
    async def get(self, endpoint: str, params: dict | None = None, headers=None) -> dict:
        if not headers:
            headers = self._generate_headers(self.token)
        await self.set_session()
        async with self.session.get(
            endpoint, params=params, headers=headers
        ) as response:
            if hasattr(self, 'populate_headers'):
                self.populate_headers(response)
            if not response.ok:
                print(response)
                raise Exception(f"Error {endpoint} {response.status}: {await response.text()}")
            if response.status == 204:
                logger.info(f"GET {endpoint} returned 204 No Content")
                return {}
            return await response.json()

    async def post(self, endpoint: str, data: dict | None = None, headers=None) -> dict:
        if not headers:
            headers = self._generate_headers()
        await self.set_session()
        async with self.session.post(
            endpoint, json=data, headers=headers
        ) as response:
            if hasattr(self, 'populate_headers'):
                self.populate_headers(response)
            if not response.ok:
                raise Exception(f"Error {response.status}: {await response.text()}")
            if response.status == 204:
                logger.info(f"POST {endpoint} returned 204 No Content")
                return {}
            return await response.json()

    async def api(self, command: Command):
        method = command.method.lower()
        if method == "get":
            coro = self.get(
                command.endpoint, 
                params=command.params
                )
        elif method == "post":
            coro = self.post(
                command.endpoint, 
                data=command.params
            )

        else:
            raise NotImplementedError(f"HTTP method {command.method} not supported.")
        response = await coro
        if command.callback:
            return command.callback(response)
        return response