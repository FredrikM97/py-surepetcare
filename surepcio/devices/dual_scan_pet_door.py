import logging

from .device import BaseControl
from .device import BaseStatus
from .device import DeviceBase
from surepcio.command import Command
from surepcio.const import API_ENDPOINT_PRODUCTION
from surepcio.devices.entities import DevicePetTag
from surepcio.enums import ProductId

logger = logging.getLogger(__name__)


class DualScanPetDoor(DeviceBase):
    def __init__(self, data: dict, **kwargs) -> None:
        try:
            super().__init__(data, **kwargs)
            self.status: BaseStatus = BaseStatus(**data)
            self.control: BaseControl = BaseControl(**data)
        except Exception as e:
            logger.warning("Error while storing data %s", data)
            raise e

    @property
    def product(self) -> ProductId:
        return ProductId.DUAL_SCAN_PET_DOOR

    def refresh(self):
        def parse(response):
            if not response:
                return self
            self.status = BaseStatus(**{**self.status.model_dump(), **response["data"]})
            self.control = BaseControl(**{**self.control.model_dump(), **response["data"]})
            self.tags = [DevicePetTag(**tag) for tag in response["data"].get("tags", [])]
            return self

        return Command(
            method="GET",
            endpoint=f"{API_ENDPOINT_PRODUCTION}/device/{self.id}",
            callback=parse,
        )
