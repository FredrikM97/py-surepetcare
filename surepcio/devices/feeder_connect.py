import logging
from datetime import datetime
from typing import Optional

from .device import BaseControl
from .device import BaseStatus
from .device import DeviceBase
from surepcio.command import Command
from surepcio.const import API_ENDPOINT_PRODUCTION
from surepcio.devices.entities import DevicePetTag
from surepcio.devices.entities import FlattenWrappersMixin
from surepcio.enums import BowlPosition
from surepcio.enums import CloseDelay
from surepcio.enums import FoodType
from surepcio.enums import ProductId

logger = logging.getLogger(__name__)


class BowlState(FlattenWrappersMixin):
    position: BowlPosition = BowlPosition.LEFT
    food_type: FoodType = FoodType.UNKNOWN
    substance_type: Optional[int] = None
    current_weight: Optional[float] = None
    last_filled_at: datetime
    last_zeroed_at: datetime
    last_fill_weight: Optional[float] = None
    fill_percent: Optional[int] = None


class BowlTargetWeight(FlattenWrappersMixin):
    food_type: FoodType = FoodType.DRY
    full_weight: Optional[int] = None


class Lid(FlattenWrappersMixin):
    close_delay: CloseDelay


class Control(BaseControl):
    lid: Optional[Lid] = None
    bowls: Optional[BowlTargetWeight] = None
    tare: Optional[int] = None
    training_mode: Optional[int] = None
    fast_polling: Optional[bool] = None


class Status(BaseStatus):
    bowl_status: Optional[list[BowlState]] = None


class FeederConnect(DeviceBase):
    def __init__(self, data: dict, **kwargs) -> None:
        try:
            super().__init__(data, **kwargs)
            self.status: Status = Status(**data)
            self.control: Control = Control(**data)
        except Exception as e:
            logger.warning("Error while storing data %s", data)
            raise e

    @property
    def product(self) -> ProductId:
        return ProductId.FEEDER_CONNECT

    @property
    def photo(self) -> str:
        return "https://www.surepetcare.io/assets/assets/products/feeder.7ff330c9e368df01d256156b6fc797bb.png"

    def refresh(self):
        def parse(response):
            if not response:
                return self
            self.status = Status(**{**self.status.model_dump(), **response["data"]})
            self.control = Control(**{**self.control.model_dump(), **response["data"]})
            self.tags = [DevicePetTag(**tag) for tag in response["data"].get("tags", [])]
            return self

        command = Command(
            method="GET",
            endpoint=f"{API_ENDPOINT_PRODUCTION}/device/{self.id}",
            callback=parse,
        )
        return command

    @property
    def rssi(self) -> Optional[int]:
        """Return the RSSI value."""
        return self.status.signal.device_rssi if self.status.signal else None
