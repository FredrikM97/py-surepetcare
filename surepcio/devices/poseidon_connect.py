import logging
from typing import Optional

from .device import BaseControl
from .device import BaseStatus
from .device import DeviceBase
from surepcio.command import Command
from surepcio.const import API_ENDPOINT_PRODUCTION
from surepcio.devices.entities import BowlState
from surepcio.enums import ProductId

logger = logging.getLogger(__name__)


class Control(BaseControl):
    learn_mode: Optional[bool] = None
    fast_polling: Optional[bool] = None


class Status(BaseStatus):
    bowl_status: Optional[list[BowlState]] = None


class PoseidonConnect(DeviceBase[Control, Status]):
    """Representation of a Poseidon Connect device."""

    controlCls = Control
    statusCls = Status

    def refresh(self):
        """Refresh the device status and control settings from the API."""

        def parse(response) -> "PoseidonConnect":
            if not response:
                return self

            self.status = self.statusCls(**{**self.status.model_dump(), **response["data"]})
            self.control = self.controlCls(**{**self.control.model_dump(), **response["data"]})
            return self

        return Command(
            method="GET",
            endpoint=f"{API_ENDPOINT_PRODUCTION}/device/{self.id}",
            callback=parse,
        )

    @property
    def product(self) -> ProductId:
        return ProductId.POSEIDON_CONNECT
