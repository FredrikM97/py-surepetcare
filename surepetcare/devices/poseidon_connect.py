from surepy.devices.base import SurepyDevice
from surepy.enums import ProductId


class PoseidonConnect(SurepyDevice):
    @property
    def product(self) -> ProductId:
        return ProductId.POSEIDON_CONNECT
