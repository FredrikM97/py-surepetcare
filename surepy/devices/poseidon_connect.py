from surepy.devices.base import SurepyDevice
from surepy.enums import ProductId

class PoseidonConnect(SurepyDevice):
    product_id:ProductId = ProductId.POSEIDON_CONNECT
