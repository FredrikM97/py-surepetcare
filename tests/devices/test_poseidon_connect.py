# Integration tests for PoseidonConnect device (add real integration if/when mock data is available)
from surepetcare.devices.poseidon_connect import PoseidonConnect
from surepetcare.enums import ProductId


class DummyClient:
    pass


def make_data():
    return {
        "id": 123,
        "household_id": 456,
        "name": "Test Device",
        "status": {"online": True, "battery": 6.0},
    }


def test_poseidon_connect_properties():
    device = PoseidonConnect(DummyClient(), make_data())
    assert device.product == ProductId.POSEIDON_CONNECT
