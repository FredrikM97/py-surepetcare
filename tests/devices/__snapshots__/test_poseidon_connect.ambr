# serializer version: 1
# name: test_snapshot[device_names0]
  dict({
    'control': dict({
      'fast_polling': False,
      'learn_mode': False,
      'tags': None,
    }),
    'entity_info': dict({
      'household_id': 7777,
      'id': 734730,
      'name': 'Poseidon connect',
      'parent_device_id': None,
      'photo': None,
      'product_id': 8,
      'tag': None,
      'tag_id': None,
    }),
    'status': dict({
      'battery': 5.514,
      'bowl_status': list([
        dict({
          'current_weight': 975.95,
          'fill_percent': 91,
          'food_type': None,
          'last_fill_weight': 1071.92,
          'last_filled_at': datetime.datetime(2025, 9, 3, 8, 39, 29, tzinfo=TzInfo(UTC)),
          'last_zeroed_at': None,
          'position': <BowlPosition.ONE: 0>,
          'substance_type': <SubstanceType.WATER: 1>,
        }),
      ]),
      'learn_mode': False,
      'online': True,
      'signal': dict({
        'device_rssi': -74,
      }),
      'version': dict({
        'device': dict({
          'firmware': '213.562',
          'hardware': '1',
        }),
      }),
    }),
    'timezone': 'Europe/Stockholm',
  })
# ---
